go.mod

```diff
@@ -1,6 +1,6 @@
 module github.com/gametimesf/aws-ssm-env/v2
 
-go 1.18
+go 1.21.5
```

Comment:
the version in `go.mod` specifies what language features are required. it does not control anything for compilation other than unnecessarily causing earlier versions to refuse to compile it.

at any rate, there is very rarely a reason to specify a minor version here.

Reply:
I tested out using 1.20 and compiling with 1.20.0 and there's still some unpatched vulnerabilities that exist that aren't patched until this minor version. The update above is to ensure anyone performing the release and updating the binaries manually will do so with the version above or higher. The vulns are within the Go stdlib itself.

Reply:
When we release these binaries we compile the build so like @ajanata mentioned it relies on what version was used to compile. I think you should add a README.md with the instructions that tells the minimum version of go to create the release with to avoid the vulnerability

Reply:
Sounds good, I can add to the README. Whoever is compiling will need to be on version 1.21 or higher, I can check for this in the Makefile as well. The change above is to enforce that it'll be compiled with that specific version of Go or higher (pretty cool 😎 ), which was introduced to be automagically done in Go 1.21. From [Forward Compatibility and Toolchain Management in Go 1.21](https://go.dev/blog/toolchain):

> When you need a new version of a Go module, the go command downloads it for you. Starting in Go 1.21, when you need a newer Go toolchain, the go command downloads that for you too. This functionality is like Node’s nvm or Rust’s rustup, but built in to the core go command instead of being a separate tool.

> If you are running Go 1.21.0 and you run a go command, say, go build, in a module with a go.mod that says go 1.21.1, the Go 1.21.0 go command will notice that you need Go 1.21.1, download it, and re-invoke that version’s go command to finish the build. When the go command downloads and runs these other toolchains, it doesn’t install them in your PATH or overwrite the current installation. Instead, it downloads them as Go modules, inheriting all the [security and privacy benefits of modules](https://go.dev/blog/module-mirror-launch), and then it runs them from the module cache.

Reply:
I still think this is better indicated with the `toolchain` directive than the `go` directive since the problem is with the implementation of the standard library, not the language specification itself.

Reply:
I tested that out but couldn't get it to work. Per the [toolchain documentation](https://go.dev/doc/toolchain):

> Before Go 1.21, the suggested way to update a module to a new Go version (say, Go 1.22) was go mod tidy -go=1.22, to make sure that any adjustments specific to Go 1.22 were made to the go.mod at the same time that the go line is updated. That form is still valid, but the simpler go get go@1.22 is now preferred.

1. Go major version with toolchain minor
```
❯ go get go@1.21 toolchain@1.21.5
go: upgraded go 1.21 => 1.21.5

-----go.mod result-----
go 1.21.5
```

2. Both minor
```
❯ go get go@1.21.0 toolchain@1.21.5
go: downgraded go 1.21.5 => 1.21.0
go: added toolchain go1.21.5

-----go.mod result-----
go 1.21.0

toolchain go1.21.5
```

3. Using the doc recommendation from above
```
❯ go get go@1.21.5
go: upgraded go 1.21.0 => 1.21.5
go: removed toolchain go1.21.5

-----go.mod result-----
go 1.21.5
```

Reply:
I think it's sufficient to just `go get toolchain@1.21.5`: that inserts the appropriate `toolchain` directive but leaves the `go` directive alone. Old versions of go will error on the unknown directive, and go 1.21 will still process it properly (`go version` is telling me 1.21.5 in my test directory, but 1.21.0 outside of it).

and if we want to be thorough, this could be one of the cases where modifying the file by hand is appropriate